daemon/src/auto_rollover.rs

```diff
@@ -112,8 +115,43 @@ where
     M: xtra::Handler<monitor::StartMonitoring>,
     O: xtra::Handler<oracle::MonitorAttestation> + xtra::Handler<oracle::GetAnnouncement>,
 {
-    async fn handle_rollover_completed(&mut self, _: RolloverCompleted) -> Result<()> {
-        // TODO: Implement this in terms of event sourcing
+    async fn handle_rollover_completed(&mut self, msg: RolloverCompleted) -> Result<()> {
+        let mut conn = self.db.acquire().await?;
+        let order_id = msg.order_id();
+
+        let cfd = load_cfd(order_id, &mut conn).await?;
+
+        let event = match cfd.rollover(msg)? {
+            Some(event) => event,
+            None => return Ok(()),
+        };
+
+        db::append_event(event.clone(), &mut conn).await?;
+
+        self.projection_actor.send(projection::CfdsChanged).await?;
+
+        let dlc = match event.event {
+            CfdEvent::RolloverCompleted { dlc } => dlc,
+            CfdEvent::RolloverFailed | CfdEvent::RolloverRejected => {
+                return Ok(());
+            }
+            _ => bail!("Unexpected event {:?}", event.event),
+        };
+
+        tracing::info!("Setup complete, publishing on chain now");
```

Comment:
This message is a copy-paste from contract-setup and incorrect in this context.

Reply:
Thanks for pointing it out!